_posts/work/2018-01-30-hdf-in-the-cloud.md

```diff
@@ -0,0 +1,311 @@
+Looking at HDF traces it looks like we can probably reduce "dozens" of web requests to "a few".
+
+FUSE also requires elevated operating system permissions,
+which can introduce challenges if working from Docker containers
+(which is likely on the cloud).
+Docker containers running FUSE need to be running in privileged mode.
+There are some tricks around this, with Kubernetes, notably FlexVolumes,
+but generally FUSE brings some excess baggage.
+
+
+HDF and a Custom Reader
+-----------------------
+
+The HDF library doesn't *need* to use C File pointers,
+we can extend it to use other storage backends as well.
+
+This has already been done to support Amazon's S3 twice,
+
+1.  Once by the [HDF group](https://www.hdfgroup.org/) (currently private private),
+2.  Once by Joe Jevnik and Scott Sanderson (Quantopian) at [https://h5s3.github.io/h5s3/](https://h5s3.github.io/h5s3/) (highly [experimental](https://h5s3.github.io/h5s3/warnings.html))
+
+This provides an alternative to FUSE.
+This is better because it doesn't require privileged access,
+but is worse because it only solves this problem for HDF and not all file access.
+
+In either case we'll need to do look-ahead buffering and caching to get reasonable performance.
+
+
+Centralize Metadata
+-------------------
+
+Alternatively, we might centralize metadata in one place so that we can avoid many hops through the file.
+This could be within the HDF format itself, or in a separate side-file.
+
+This would remove the need to perform clever file-system caching and buffering tricks,
+but requires some work on the HDF library itself.
+
+
+HDF + Web Service
+-----------------
+
+We could offload this problem to a for-profit entity,
+like the [HDF group](https://www.hdfgroup.org/) or a cloud provider (Google, Amazon, Microsoft).
+They would solve this problem however they like,
+and expose a web API that we can hit for our data.
+
+This would be distributed service of several computers on the cloud near our data
+that will take our requests for what data we want,
+perform whatever tricks they deem appropriate to get that data,
+and then deliver it to us.
+This fleet of machines will still have to address the problems listed above,
+but we can let them figure it out, and presumably they'll learn as they go.
+
+However, this is somewhat complex, and creates an intermediary between us and our data.
+
+The HDF group is working on such a service, HSDS that works on Amazon's S3 (or anything that looks like S3).
+They have created a [h5pyd](https://github.com/HDFGroup/h5pyd) library that is a drop-in replacement for the popular [h5py](https://github.com/HDFGroup/h5pyd) Python library.
+
+Presumably a cloud provider, like Amazon, Google, or Microsoft could do this as well.
+By providing open standards like [OpenDAP](https://www.opendap.org/) they might attract more science users onto their platform
+to more efficiently query their cloud-hosted datasets.
```

Comment:
Maybe worth mentioning that some companies already provide similar APIs for their cloud climate data stores, e.g.
https://medium.com/planet-os/an-api-for-the-worlds-weather-climate-data-7e9946169f48